### PR TITLE
Record voltages

### PR DESCRIPTION
**Pull request by mundya**
Allows recording of neuron voltages.

After review I'd like to squash before merging.

### CODE REVIEW

**Comment by mundya** [comment]
@neworderofjamie, can I ask for a review of this please? It appears my fears of merging are reducing.

File: spinnaker_components/ensemble/recording.c [hunk 1]
```diff
@@ -24,3 +24,49 @@ void record_buffer_reset(recording_buffer_t *buffer)
...
+bool record_buffer_initialise_spikes(
+  recording_buffer_t *buffer,
+  address_t region,
+  uint n_blocks,
+  uint n_neurons
+)
+{
+  // Compute the block length (an integral number of words allowing for 1 bit
+  // per neuron).
+  uint32_t block_length = (n_neurons >> 5) + (n_neurons & 0x1f ? 1 : 0);
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.c]: n_neurons / 32 will compile to the same and make the meaning clearer

File: spinnaker_components/ensemble/recording.c [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+{
+  // Store buffer parameters
   buffer->n_blocks = n_blocks;
-  buffer->_sdram_start = (uint *) region;
+  buffer->block_length = block_length;
+  buffer->_sdram_start = (uint32_t *) region;
   record_buffer_reset(buffer);
 
   // Create the local buffer
   MALLOC_FAIL_FALSE(buffer->buffer, buffer->block_length * sizeof(uint));
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.c]: sizeof(uint32_t)

File: spinnaker_components/ensemble/recording.c [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
 bool record_buffer_initialise(recording_buffer_t *buffer, address_t region,
-                              uint n_blocks, uint n_neurons) {
-  // Generate and store buffer parameters
-  buffer->block_length = (n_neurons >> 5) + (n_neurons & 0x1f ? 1 : 0);
+                              uint32_t n_blocks, uint32_t block_length)
+{
+  // Store buffer parameters
   buffer->n_blocks = n_blocks;
-  buffer->_sdram_start = (uint *) region;
+  buffer->block_length = block_length;
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.c]: I'd have assumed this was in bytes if I hadn't read further down - rename to block_length_words, block_words or something?

File: spinnaker_components/ensemble/recording.c [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
+  buffer->_sdram_start = (uint32_t *) region;
   record_buffer_reset(buffer);
 
   // Create the local buffer
   MALLOC_FAIL_FALSE(buffer->buffer, buffer->block_length * sizeof(uint));
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.c]: sizeof(uint32_t)

File: spinnaker_components/ensemble/recording.c [hunk 5]
```diff
@@ -24,3 +24,49 @@ void record_buffer_reset(recording_buffer_t *buffer)
...
+bool record_buffer_initialise_voltages(
+  recording_buffer_t *buffer,
+  address_t region,
+  uint n_blocks,
+  uint n_neurons
+)
+{
+  // Compute the block length. We allow for 1 short per neuron and then round
+  // up to an integral number of words.
+  uint32_t block_length = (n_neurons >> 1) + (n_neurons & 0x1);
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.c]: Again / 2 and block_words would make this clearer

File: spinnaker_components/ensemble/recording.h [hunk 6]
```diff
@@ -15,56 +15,97 @@
...
+/*!\brief Record a voltage for the given neuron.
+ */
+static inline void record_voltage(
+    recording_buffer_t *buffer, uint32_t n_neuron, value_t voltage
+)
+{
+  // Cast the recording buffer to U1.15 and then store the voltage in this
+  // form.
+  uint16_t *data = (uint16_t *) buffer->buffer;
+  data[n_neuron] = bitsk(voltage) & 0xffff;
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/recording.h]: I'd contemplate using a union{ accum; uint16_t[2] } style construction here. Exactly what sign extending/casting occurs here is not immediately obvious

**Comment by neworderofjamie** [comment]
Rigged

**Comment by mundya** [comment]
Danke

**Pull request merged**